src/us/kbase/test/auth2/service/ui/AdminTest.java

```diff
@@ -0,0 +1,236 @@
+								ConfigItem.state(true),
+								ConfigItem.state(true)),
+						52000));
+		
+		assertThat("incorrect config", admin.getConfig(headers, uriInfo), is(MapBuilder.newHashMap()
+				.with("updatetimesec", 52)
+				.with("allowlogin", true)
+				.with("allowedloginredirect", new URL("http://u1.com"))
+				.with("completeloginredirect", new URL("http://u2.com"))
+				.with("completelinkredirect", new URL("http://u4.com"))
+				.with("postlinkredirect", new URL("http://u3.com"))
+				.with("ignoreip", true)
+				.with("showstack", true)
+				.with("tokenlogin", 84L)
+				.with("tokenagent", 8L)
+				.with("tokendev", 42L)
+				.with("tokenserv", 2L)
+				.with("tokensugcache", 7L)
+				.with("providers", Arrays.asList(
+						ImmutableMap.of(
+								"provider", "prov1",
+								"enabled", false,
+								"forceloginchoice", false,
+								"forcelinkchoice", false),
+						ImmutableMap.of(
+								"provider", "prov2",
+								"enabled", true,
+								"forceloginchoice", true,
+								"forcelinkchoice", true)
+						))
+				.with("tokenurl", "")
+				.with("cfgbasicurl", "../basic")
+				.with("providerurl", "../provider")
+				.with("reseturl", "../reset")
+				.build()));
+	}
+	
+	@Test
+	public void getConfigFailNoTokenProvided() {
+		final Authentication auth = mock(Authentication.class);
+		final AuthAPIStaticConfig cfg = new AuthAPIStaticConfig("kbcookie");
+		final HttpHeaders headers = mock(HttpHeaders.class);
+		final UriInfo uriInfo = mock(UriInfo.class);
+		
+		final Admin admin = new Admin(auth, cfg);
+		
+		when(headers.getCookies()).thenReturn(
+				ImmutableMap.of("kbcookie2", new Cookie("kbcookie", "token")));
+		
+		failGetConfig(admin, headers, uriInfo,
+				new NoTokenProvidedException("No user token provided"));
+	}
+	
+	@Test
+	public void getConfigFailMapping() throws Exception {
+		final Authentication auth = mock(Authentication.class);
+		final AuthAPIStaticConfig cfg = new AuthAPIStaticConfig("kbcookie");
+		final HttpHeaders headers = mock(HttpHeaders.class);
+		final UriInfo uriInfo = mock(UriInfo.class);
+		
+		final Admin admin = new Admin(auth, cfg);
+		
+		when(headers.getCookies()).thenReturn(
+				ImmutableMap.of("kbcookie", new Cookie("kbcookie", "token")));
+		
+		when(auth.getConfig(eq(new IncomingToken("token")), any(AuthExternalConfigMapper.class)))
+				.thenThrow(new ExternalConfigMappingException("foo"));
+		
+		failGetConfig(admin, headers, uriInfo,
+				new RuntimeException("There's something very wrong in the database config"));
+	}
+	
+	@Test
+	public void getConfigFailUnauthorized() throws Exception {
+		final Authentication auth = mock(Authentication.class);
+		final AuthAPIStaticConfig cfg = new AuthAPIStaticConfig("kbcookie");
+		final HttpHeaders headers = mock(HttpHeaders.class);
+		final UriInfo uriInfo = mock(UriInfo.class);
+		
+		final Admin admin = new Admin(auth, cfg);
+		
+		when(headers.getCookies()).thenReturn(
+				ImmutableMap.of("kbcookie", new Cookie("kbcookie", "token")));
```

Comment:
Seems like it would be cleaner to isolate the mocks that are invariant into a setup function but I don't know if would cause problems

Reply:
I'm of two minds about this. I used to do it religiously to DRY the tests, but I found for smaller set ups like this that it was much easier for me to read the tests if all the set up was right there in the test body and I didn't have to jump around to figure out which stubs were set up. For more significant set ups it's definitely worth it; the alternative is a bunch of bloated tests. Here I'm not so sure.

Also none of the mocks are invariant actually; they're all modified in at least one test.

I don't feel particularly strongly about this point, so if you do I'm happy to do it in a future PR - let me know.

Reply:
I don't feel strongly about this so let's proceed.